### PR TITLE
Flatten the render tree into a list + Assimp material loading + Line batching

### PR DESCRIPTION
**Pull request by IronicallySerious**
One of the last times the render system will get a face lift

### CODE REVIEW

**Comment by sin3point14** [comment]
Also, reduce shader binds to once per frame

File: rootex/core/renderer/material_library.h [hunk 1]
```diff
@@ -8,22 +8,23 @@ class MaterialLibrary
...
 	typedef HashMap<String, Pair<MaterialDefaultCreator, MaterialCreator>> MaterialDatabase;
+
+	static MaterialMap s_Materials;
 	static MaterialDatabase s_MaterialDatabase;
+
 public:
+	static void SaveAll();
 	static void LoadMaterials();
+	static void CreateNewMaterialFile(const String& materialName, const String& materialType);
+	static bool IsExists(const String& materialName);
```
[the remark this answers is not in the document]
Comment (IronicallySerious) [reply rootex/core/renderer/material_library.h:24]: We can't predict this right now

File: rootex/core/renderer/texture.cpp [hunk 2]
```diff
@@ -1,12 +1,19 @@
...
-	
+	loadTexture();
+}
+
+Texture::Texture(uint8_t* imageData, size_t size)
+    : m_ImageFile(nullptr)
+{
+	m_TextureView = RenderingDevice::GetSingleton()->createTexture(imageData, size);
+
 	Microsoft::WRL::ComPtr<ID3D11Resource> res;
```
[the remark this answers is not in the document]
Comment (IronicallySerious) [reply rootex/core/renderer/texture.cpp:17]: No need, load texture only loads the texture for when the image file is specified. If the texture is made from raw data, image file is null

File: rootex/framework/systems/physics_system.cpp [hunk 3]
```diff
@@ -127,11 +128,15 @@ void PhysicsSystem::InternalTickCallback(btDynamicsWorld* const world, btScalar
 
 void PhysicsSystem::debugDraw()
```
[the remark this answers is not in the document]
Comment (IronicallySerious) [reply rootex/framework/systems/physics_system.cpp:129]: We will need this for debugging inside the game

File: rootex/framework/systems/render_system.cpp [hunk 4]
```diff
@@ -1,11 +1,9 @@
 #include "render_system.h"
```
[the remark this answers is not in the document]
Comment (IronicallySerious) [reply rootex/framework/systems/render_system.cpp:1]: Done

File: rootex/framework/systems/render_system.cpp [hunk 5]
```diff
@@ -22,36 +20,40 @@ RenderSystem::RenderSystem()
...
-void RenderSystem::renderPassRender(VisualComponent* vc, const RenderPass& renderPass)
+void RenderSystem::renderPassRender(RenderPass renderPass)
 {
-	vc->preRender();
-	if (vc->isVisible())
+	ModelComponent* mc = nullptr;
+	for (auto& component : s_Components[ModelComponent::s_ID])
 	{
-		vc->render(renderPass);
+		mc = (ModelComponent*)component;
```
[the remark this answers is not in the document]
Comment (IronicallySerious) [reply rootex/framework/systems/render_system.cpp:46]: For performance when we are sure that it cannot fail, yes

File: rootex/framework/systems/render_system.h [hunk 6]
```diff
@@ -1,17 +1,30 @@
...
 
+#include "core/renderer/renderer.h"
 #include "framework/components/visual/camera_component.h"
 #include "framework/system.h"
 #include "framework/systems/hierarchy_system.h"
 #include "main/window.h"
+#include "components/visual/model_component.h"
+#include "renderer/render_pass.h"
+
+#define LINE_INTIAL_RENDER_CACHE 1000
```
[the remark this answers is not in the document]
Comment (IronicallySerious) [reply rootex/framework/systems/render_system.h:11]: Done

**Pull request merged**